builtin/logical/database/dbplugin/databasemiddleware.go

```diff
@@ -0,0 +1,172 @@
+package dbplugin
+
+import (
+	"time"
+
+	metrics "github.com/armon/go-metrics"
+	log "github.com/mgutz/logxi/v1"
+)
+
+// ---- Tracing Middleware Domain ----
+
+// databaseTracingMiddleware wraps a implementation of DatabaseType and executes
+// trace logging on function call.
+type databaseTracingMiddleware struct {
+	next   DatabaseType
+	logger log.Logger
+
+	typeStr string
+}
+
+func (mw *databaseTracingMiddleware) Type() (string, error) {
+	return mw.next.Type()
+}
+
+func (mw *databaseTracingMiddleware) CreateUser(statements Statements, usernamePrefix string, expiration time.Time) (username string, password string, err error) {
+	if mw.logger.IsTrace() {
+		defer func(then time.Time) {
+			mw.logger.Trace("database/CreateUser: finished", "type", mw.typeStr, "err", err, "took", time.Since(then))
+		}(time.Now())
+
+		mw.logger.Trace("database/CreateUser: starting", "type", mw.typeStr)
+	}
+	return mw.next.CreateUser(statements, usernamePrefix, expiration)
+}
+
+func (mw *databaseTracingMiddleware) RenewUser(statements Statements, username string, expiration time.Time) (err error) {
+	if mw.logger.IsTrace() {
+		defer func(then time.Time) {
+			mw.logger.Trace("database/RenewUser: finished", "type", mw.typeStr, "err", err, "took", time.Since(then))
+		}(time.Now())
+
+		mw.logger.Trace("database/RenewUser: starting", "type", mw.typeStr)
+	}
+	return mw.next.RenewUser(statements, username, expiration)
+}
+
+func (mw *databaseTracingMiddleware) RevokeUser(statements Statements, username string) (err error) {
+	if mw.logger.IsTrace() {
+		defer func(then time.Time) {
+			mw.logger.Trace("database/RevokeUser: finished", "type", mw.typeStr, "err", err, "took", time.Since(then))
+		}(time.Now())
+
+		mw.logger.Trace("database/RevokeUser: starting", "type", mw.typeStr)
+	}
+	return mw.next.RevokeUser(statements, username)
+}
+
+func (mw *databaseTracingMiddleware) Initialize(conf map[string]interface{}, verifyConnection bool) (err error) {
+	if mw.logger.IsTrace() {
+		defer func(then time.Time) {
+			mw.logger.Trace("database/Initialize: finished", "type", mw.typeStr, "verify", verifyConnection, "err", err, "took", time.Since(then))
+		}(time.Now())
+
+		mw.logger.Trace("database/Initialize: starting", "type", mw.typeStr)
+	}
+	return mw.next.Initialize(conf, verifyConnection)
+}
+
+func (mw *databaseTracingMiddleware) Close() (err error) {
+	if mw.logger.IsTrace() {
+		defer func(then time.Time) {
+			mw.logger.Trace("database/Close: finished", "type", mw.typeStr, "err", err, "took", time.Since(then))
+		}(time.Now())
+
+		mw.logger.Trace("database/Close: starting", "type", mw.typeStr)
+	}
+	return mw.next.Close()
+}
+
+// ---- Metrics Middleware Domain ----
+
+// databaseMetricsMiddleware wraps an implementation of DatabaseTypes and on
+// function call logs metrics about this instance.
+type databaseMetricsMiddleware struct {
```

Comment:
Can this be combined with the above? e.g. a bool or flag that controls whether it should do log tracing, metrics tracing, or both, and only defers single functions?